scripts/local_cloudbuild.py

```diff
@@ -0,0 +1,322 @@
+Step = collections.namedtuple('Step', 'args dir_ env name')
+
+
+def get_field_value(container, field_name, field_type):
+    """Fetch a field from a container with typechecking and default values.
+
+    The field value is coerced to the desired type.  If the field is
+    not present, a instance of `field_type` is constructed with no
+    arguments and used as the default value.
+
+    Args:
+        container (dict): Object decoded from yaml
+        field_name (str): Field that should be present in `container`
+        field_type (type): Expected type for field value
+
+    Returns:
+        any: Fetched or default value of field
+
+    Raises:
+        ValueError: if field value cannot be converted to the desired type
+    """
+    try:
+        value = container[field_name]
+    except (IndexError, KeyError):
+        return field_type()
+
+    msg = 'Expected "{}" field to be of type "{}", but found type "{}"'
+    if not isinstance(value, field_type):
+        # list('some string') is a successful type cast as far as Python
+        # is concerned, but doesn't exactly produce the results we want.
+        # We have a whitelist of conversions we will attempt.
+        whitelist = (
+            (float, str),
+            (int, str),
+            (str, float),
+            (str, int),
+            (int, float),
+            )
+        if (type(value), field_type) not in whitelist:
+            raise ValueError(msg.format(field_name, field_type, type(value)))
+
+    try:
+        value = field_type(value)
+    except ValueError as e:
+        e.message = msg.format(field_name, field_type, type(value))
+        raise
+    return value
+
+
+def get_cloudbuild(raw_config, args):
+    """Read and validate a cloudbuild recipe
+
+    Args:
+        raw_config (dict): deserialized cloudbuild.yaml
+        args (argparse.Namespace): ccommand line flags
+
+    Returns:
+        CloudBuild: valid configuration
+    """
+    if not isinstance(raw_config, dict):
+        raise ValueError(
+            'Expected {} contents to be of type "dict", but found type "{}"'.
+            format(args.config, type(raw_config)))
+    raw_steps = get_field_value(raw_config, 'steps', list)
+    if not raw_steps:
+        raise ValueError('No steps defined in {}'.format(args.config))
+    steps = [get_step(raw_step) for raw_step in raw_steps]
+    return CloudBuild(
+        output_script=args.output_script,
+        run=args.run,
+        steps=steps,
+    )
+
+
+def get_step(raw_step):
+    """Read and validate a single cloudbuild step
+
+    Args:
+        raw_step (dict): deserialized step
+
+    Returns:
+        Step: valid build step
+    """
+    if not isinstance(raw_step, dict):
+        raise ValueError(
+            'Expected step to be of type "dict", but found type "{}"'.
+            format(type(raw_step)))
+    raw_args = get_field_value(raw_step, 'args', list)
+    args = [get_field_value(raw_args, i, str)
```

Comment:
i,j,k have a rich history of use as array indices, whereas 'n' is a total or goal number.  So 'i' is more appropriate than 'n' here.  But I just changed it to 'index' since, well, I'm not writing Fortran any more :)